internal/services/cio/cio.go

```diff
@@ -0,0 +1,83 @@
+package cio
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	pb_accounts "github.com/DIMO-Network/accounts-api/pkg/grpc"
+	"github.com/DIMO-Network/devices-api/models"
+	analytics "github.com/customerio/cdp-analytics-go"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+)
+
+const SoftwareConnectionExpiredEvent = "software_connection_expired"
+
+type AccountsClient interface {
+	GetAccount(context.Context, *pb_accounts.GetAccountRequest, ...grpc.CallOption) (*pb_accounts.Account, error)
+}
+
+type Service struct {
+	client     analytics.Client
+	logger     *zerolog.Logger
+	acctClient AccountsClient
+}
+
+func New(cioKey string, acctClient AccountsClient, logger *zerolog.Logger) (*Service, error) {
+	client, err := analytics.NewWithConfig(cioKey, analytics.Config{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		client:     client,
+		acctClient: acctClient,
+		logger:     logger,
+	}, nil
+
+}
+
+func (s *Service) SoftwareDisconnectionEvent(ctx context.Context, udai *models.UserDeviceAPIIntegration) error {
+	vehicleTokenID, ok := udai.R.UserDevice.TokenID.Int64()
+	if !ok {
+		return errors.New("failed to parse vehicle token id")
+	}
+
+	userAddr := common.BytesToAddress(udai.R.UserDevice.OwnerAddress.Bytes)
+	if !common.IsHexAddress(userAddr.Hex()) {
+		return fmt.Errorf("invalid ethereum_address %s", userAddr.Hex())
+	}
```

Comment:
I think there's no danger of this happening. If you wanted to be really, really careful you could check that `udai.R.UserDevice.OwnerAddress.Bytes` really has length 20, but I think the database will enforce that if it's populated so I think it's not too scary.

Reply:
okay cool, removed it in that case